File: week2/hw1/java/HashTable.java
package week2.hw1.java;

public class HashTable {
    private int bucketSize;
    private Item[] buckets;
    private int itemCount;

    public int calculateHash(String key){
        assert (key instanceof String);
        int hash = 1;
        for(char c : key.toCharArray()){
            hash = (int)(((long)hash * (long)c) % 1000000009L);
        }
        return hash;
    }

    public HashTable(){
        this.bucketSize = 97;
        this.buckets = new Item[bucketSize];
        this.itemCount = 0;
    }

    public void rehash(int newSize, int depth){
        Item[] oldBuckets = this.buckets;
        this.buckets = new Item[newSize];
        this.bucketSize = newSize;
        this.itemCount = 0;

        for(Item item :oldBuckets){
            while(item != null){
                this.put(item.getKey(), item.getValue(),depth+1 );
                item = item.getNext();
            }
        }

        return;
    }

    public void checkSize(int depth){
        if(this.itemCount <= this.bucketSize * 0.3){
            this.rehash(this.bucketSize / 2, depth+1);
        } else if (this.itemCount >= this.bucketSize * 0.7){
            this.rehash(this.bucketSize * 2,depth+1);
        }

        return;
    }

    public boolean put(String key, int value, int depth){
        // System.out.println(depth);
        // System.out.println("bucketSize"+this.bucketSize);
        // System.out.println("itemCnt"+this.itemCount);
        assert key instanceof String; //カッコなしでもいけるんかーい
       
        int bucketIdx = calculateHash(key) % this.bucketSize;
        Item item = this.buckets[bucketIdx];

        // System.out.println("Whileの前");
        while (item != null){
            // System.out.println(item + "をチェック中");
            if(item.getKey() == key){
                item.setValue(value);
                return false;
            }
            item = item.getNext();
        }
        // System.out.println("whileの後");
        Item newItem = new Item(key, value, this.buckets[bucketIdx]);
        this.buckets[bucketIdx] = newItem;
        this.itemCount ++;

        if (this.itemCount >= this.bucketSize * 0.7){
            this.rehash(this.bucketSize*2,depth+1);
        }
        

        return true;
    }

    public int get(String key){
        assert key instanceof String;
        int bucketIdx = calculateHash(key) % this.bucketSize;
        Item item = this.buckets[bucketIdx];

        while (item != null){
            return (item.getValue());
        }

        return -1;
    }

    public boolean delete(String key, int depth){
        assert key instanceof String;

        int bucketIdx = calculateHash(key) % this.bucketSize;
        Item item = this.buckets[bucketIdx];
        Item prev = null;

        while (item != null){
            // System.out.println("います");
            if(item.getKey().equals(key)){
                // System.out.println("います");
                if(prev == null){
                    this.buckets[bucketIdx] = item.getNext();
                } else {
                    prev.setNext(item.getNext());
                }
                this.itemCount --;

                if(this.itemCount <= this.bucketSize * 0.3){
                    this.rehash(this.bucketSize/2,depth+1);
                }
                return true;
            }
            prev = item;
            item = item.getNext();
        }
        return false;
    }

    public int size(){
        return this.itemCount;
    }

    public static void main(String[] args) {
        
    }
}

File: week2/hw4/hash_table.py
import random, sys, time

# Hash function.
#
# |key|: string
# Return value: a hash value
def calculate_hash(key):
    assert type(key) == str
    # Note: This is not a good hash function. Do you see why?
    hash = 1
    for i in key:
        # 💡 += を *=にしたら、衝突が減ったのかびっくりするくらい早くなった！！！！！
        hash = hash * ord(i) % 1000000009
    return hash


# An item object that represents one key - value pair in the hash table.
class Item:
    def __init__(self, key, value, next):
        assert type(key) == str
        self.key = key
        self.value = value
        self.next = next

class HashTable:

    # Initialize the hash table.
    def __init__(self):
        # Set the initial bucket size to 97. A prime number is chosen to reduce
        # hash conflicts.
        self.bucket_size = 97
        self.buckets = [None] * self.bucket_size
        self.item_count = 0

    # 💡 再ハッシュの実装
    def rehash(self,new_bucket_size):
        old_buckets = self.buckets
        self.buckets = [None] * new_bucket_size #
        old_bucket_size = self.bucket_size
        self.bucket_size = new_bucket_size
        self.item_count = 0

        for head in old_buckets:
            item = head
            while item:
                self.put(item.key, item.value)
                item = item.next
    
    def put(self, key, value):
        assert type(key) == str
        self.check_size() # Note: Don't remove this code.
        bucket_index = calculate_hash(key) % self.bucket_size
        item = self.buckets[bucket_index]
        while item:
            if item.key == key:
                item.value = value
                return False
            item = item.next
        new_item = Item(key, value, self.buckets[bucket_index])
        self.buckets[bucket_index] = new_item
        self.item_count += 1

        # 💡 再ハッシュの確認
        if self.item_count >= self.bucket_size * 0.7:
            self.rehash(self.bucket_size * 2)

        return True

    def get(self, key):
        assert type(key) == str
        #  ❓ これ何👇
        self.check_size() # Note: Don't remove this code. <- ❓ これ何
        bucket_index = calculate_hash(key) % self.bucket_size
        item = self.buckets[bucket_index]
        while item:
            if item.key == key:
                return (item.value, True)
            item = item.next
        return (None, False)


    def delete(self, key):
        assert type(key) == str

        self.check_size()
        bucket_index = calculate_hash(key) % self.bucket_size
        item = self.buckets[bucket_index]
        prev = None

        while item:
            if item.key == key:
                if prev is None:
                    self.buckets[bucket_index] = item.next
                else:
                    prev.next = item.next
                self.item_count -= 1

                # 💡 再ハッシュの確認
                if self.item_count <= self.bucket_size * 0.3:
                    self.rehash(self.bucket_size // 2)
                return True
            prev = item
            item = item.next
        return (False)
    
    # 💡 追加で実装、ほぼgetと同じ！
    def contains_key(self,key):
        assert type(key) == str
        bucket_index = calculate_hash(key) % self.bucket_size
        item = self.buckets[bucket_index]
        while item:
            if item.key == key:
                return True
            item = item.next
        return False
    
    # 💡 こっちも追加！バケットごとにみてvalueをappendした
    def get_value_array(self):
        values = []
        for b in self.buckets:
            item = b
            while item:
                values.append(item.value)
                item = item.next

        return values
    
    # 💡 こっちも追加！
    def get_key_array(self):
        keys = []
        for b in self.buckets:
            item = b
            while item:
                keys.append(item.key)
                item = item.next

        return keys

    # Return the total number of items in the hash table.
    def size(self):
        return self.item_count

    def check_size(self):
        pass
   